### ISSUE TITLE
Class-path entry lead to heavy download

### ISSUE THREAD

**Issue by SaiyanRiku** [bytedeco/javacpp#150]
Hi,

i have a project using a java applet which integrate javacpp and ffmpeg wrapper with all native library.

The manifest.mf of the is like this ffmpeg-3.1.2-1.2.jar :
> Manifest-Version: 1.0
Implementation-Vendor: Bytedeco
Implementation-Title: JavaCPP Presets for FFmpeg
Implementation-Version: 3.1.2-1.2
Built-By: root
Build-Jdk: 1.7.0_101
Class-Path: . javacpp.jar ffmpeg-linux-arm.jar ffmpeg-linux-x86.jar ff
 mpeg-linux-x86_64.jar ffmpeg-macosx-x86.jar ffmpeg-macosx-x86_64.jar 
 ffmpeg-windows-x86.jar ffmpeg-windows-x86_64.jar javacpp-1.2.jar
Specification-Vendor: Bytedeco
Name: org/bytedeco/javacpp/presets/
Specification-Title: JavaCPP Presets for FFmpeg
Created-By: Apache Maven 3.0.5
Specification-Version: 3.1.2-1.2
Archiver-Version: Plexus Archiver

The problem is that at the startup the applet is downloading all entry present is the class-path which can be long very on slow connection.

I fixed it with a script overriding the Class-Path, but i think it should be better if the released version doesn't add the native jar in the entry, since it's more easy to add them that remove them.

Note that the codebase_lookup=false parameters for applet doesn't resolve the problem.

**Comment by saudet** [comment]
I'm not sure I'm following you. Applets don't look in the manifest file for JAR files to download. What tools are you using?

**Comment by saudet** [comment]
In any case, in order to run applet anywhere these days, we need to provide signed JAR files:
https://github.com/bytedeco/javacv/wiki/How-To-Make-an-Applet
So, before signing your JAR files, we can modify the Class-Path in anyway you want.

**Comment by SaiyanRiku** [comment]
I know that is not a blocking problem, but it's not easy to understand how to fix the problem.

Note that i'm using icedtea-plugin (1.5.3-1) on Debian Jessie.

In my example i have this html code for applet :  

```
<applet codebase="http://host:port/resource/applets/signed/" code="Main" width="800" height="600" >
  <param name="archive" value="myapplet.jar,javacpp.jar,ffmpeg.jar,ffmpeg-windows-x86.jar" />
 <param name="codebase_lookup" value="false" />
 <param name="separate_jvm" value="true" />
 <param name="archive" value="myapplet.jar" />
 No JAVA.
 </applet>
```

This result is this set of request : 

```
HEAD /resource/applets/signed/myapplet.jar
GET /resource/applets/signed/myapplet.jar
HEAD /resource/applets/signed/ffmpeg.jar?version-id=1
GET /resource/applets/signed/ffmpeg.jar?version-id=1
HEAD /resource/applets/signed/ffmpeg-windows-x86.jar?version-id=1
GET /resource/applets/signed/ffmpeg-windows-x86.jar?version-id=1
HEAD /resource/applets/signed/javacpp.jar?version-id=1
GET /resource/applets/signed/javacpp.jar?version-id=1
GET /resource/applets/signed/ffmpeg.jar?version-id=1
GET /resource/applets/signed/myapplet.jar
GET /resource/applets/signed/ffmpeg-windows-x86.jar?version-id=1
GET /resource/applets/signed/javacpp.jar?version-id=1
HEAD /resource/applets/signed/ffmpeg-macosx-x86_64.jar
GET /resource/applets/signed/ffmpeg-macosx-x86_64.jar
GET /resource/applets/signed/ffmpeg-macosx-x86_64.jar
HEAD /resource/applets/signed/ffmpeg-linux-x86.jar
GET /resource/applets/signed/ffmpeg-linux-x86.jar
GET /resource/applets/signed/ffmpeg-linux-x86.jar
HEAD /resource/applets/signed/ffmpeg-macosx-x86.jar
GET /resource/applets/signed/ffmpeg-macosx-x86.jar
Writing HTTP exception (404 => Not Found) for request
GET /resource/applets/signed/ffmpeg-macosx-x86.jar
Writing HTTP exception (404 => Not Found) for request
HEAD /resource/applets/signed/ffmpeg-windows-x86_64.jar
GET /resource/applets/signed/ffmpeg-windows-x86_64.jar
GET /resource/applets/signed/ffmpeg-windows-x86_64.jar
HEAD /resource/applets/signed/.
GET /resource/applets/signed/.
Writing HTTP exception (400 => Bad Request) for request
HEAD /resource/applets/signed/ffmpeg-linux-x86_64.jar
GET /resource/applets/signed/ffmpeg-linux-x86_64.jar
GET /resource/applets/signed/ffmpeg-linux-x86_64.jar
HEAD /resource/applets/signed/javacpp-1.2.jar
GET /resource/applets/signed/javacpp-1.2.jar
Writing HTTP exception (404 => Not Found) for request
GET /resource/applets/signed/javacpp-1.2.jar
Writing HTTP exception (404 => Not Found) for request
HEAD /resource/applets/signed/ffmpeg-linux-arm.jar
GET /resource/applets/signed/ffmpeg-linux-arm.jar
Writing HTTP exception (404 => Not Found) for request
GET /resource/applets/signed/ffmpeg-linux-arm.jar
Writing HTTP exception (404 => Not Found) for request
GET /resource/applets/signed/myapplet.jar
```

With java sun on Windows, i got some unwanted request for example : 
`GET /resource/applets/signed/org/bytedeco/javacpp/avcodec$AVCodecContext.class`


To fix this i have to : 
- change the Manifest Class-Path of the ffmpeg.jar
- create a jar index within myappler.jar/META_INF for myapplet.jar, javacpp.jar, ffmpeg.jar (but not the native library jars)
- use the cache_archive and cache_version parameters for the applet html tag. For example :
```
      <param name="cache_archive" value="myapplet.jar,javacpp.jar,ffmpeg.jar,ffmpeg-windows-x86.jar" />
      <param name="cache_version" value="1,1,1,1" />
```

I just want to share my experience when integrating javacpp with ffmpeg in a java applet. But i'm just thinking that ffmpeg.jar should not include all native deps in it's class-path.

Then you decide if it's a bug or not.

**Comment by saudet** [comment]
Well, it's not a bug, but unless there's good reasons to change the behavior, I'd rather not change things that are currently working for desktop users. 

I see you're using the applet tag. If you don't need to use the applet tag, maybe you could try to launch the applet via JNLP? That has a lot more options...

In any case, please feel free to update the Wiki page about applets with this information:
https://github.com/bytedeco/javacv/wiki/How-To-Make-an-Applet
That's what it's there for!

**Comment by saudet** [comment]
I just an an idea. I'll remove the list of dependencies from those JAR files, but will be adding them to the new "platform artifacts". So, for example, `ffmpeg.jar` will only depend on `javacpp.jar`, but `ffmpeg-platform.jar` will be a dummy artifact with a manifest file that depends on the binaries for all platforms. This will give users the choice!

**Comment by SaiyanRiku** [comment]
Actually, my application works with JNLP, but i just debug it with the applet tag to understand the problem and explain the base of my problem.

I'll will try later to make and share a sample project to demonstrate applet usage with ffmpeg.

Yes, sure your idea will work better for me, but i understand that for desktop user it will add one dependencies more and may break compatibility.

**Comment by saudet** [comment]
Ok, the "Class-Path" entries are now clean starting with the just released version 1.3. Enjoy!

**Comment by SaiyanRiku** [comment]
Thank you very much :)
